ambari-server/src/main/java/org/apache/ambari/server/checks/DatabaseConsistencyCheckHelper.java

```diff
@@ -1147,6 +1159,248 @@ static void checkServiceConfigs()  {
+                  Set<Long> hostIds = nonMappedHostIds.get(configGroup.getId());
+                  if (CollectionUtils.isEmpty(hostIds)) {
+                    hostIds = new HashSet<>();
+                    nonMappedHostIds.put(configGroup.getId(), hostIds);
+                  }
+                  hostIds.add(host.getHostId());
+                  hostnames.add(host.getHostName());
+                  addToOutput = true;
+                }
+              }
+            }
+            if (addToOutput) {
+              output.append("( ");
+              output.append(configGroup.getName());
+              output.append(", ");
+              output.append(configGroup.getTag());
+              output.append(", ");
+              output.append(hostnames);
+              output.append(" ), ");
+            }
+          }
+        }
+      }
+    }
+    if (!MapUtils.isEmpty(nonMappedHostIds) && warnIfFound) {
+      output.replace(output.lastIndexOf(","), output.length(), "]");
+      warning("You have config group host mappings with hosts that are no " +
+        "longer associated with the cluster, {}. Run --auto-fix-database to " +
+        "fix this automatically. Alternatively, you can remove this mapping " +
+        "from the UI. Please backup Ambari Server database before running --auto-fix-database.", output.toString());
+    }
+
+    return nonMappedHostIds;
+  }
+
+  static Map<Long, ConfigGroup> checkConfigGroupsForDeletedServices(boolean warnIfFound) {
+    Map<Long, ConfigGroup> configGroupMap = new HashMap<>();
+    Clusters clusters = injector.getInstance(Clusters.class);
+    Map<String, Cluster> clusterMap = clusters.getClusters();
+    StringBuilder output = new StringBuilder("[(ConfigGroup, Service) => ");
+
+    if (!MapUtils.isEmpty(clusterMap)) {
+      for (Cluster cluster : clusterMap.values()) {
+        Map<Long, ConfigGroup> configGroups = cluster.getConfigGroups();
+        Map<String, Service> services = cluster.getServices();
+
+        if (!MapUtils.isEmpty(configGroups)) {
+          for (ConfigGroup configGroup : configGroups.values()) {
+            if (!services.containsKey(configGroup.getServiceName())) {
+              configGroupMap.put(configGroup.getId(), configGroup);
+              output.append("( ");
+              output.append(configGroup.getName());
+              output.append(", ");
+              output.append(configGroup.getServiceName());
+              output.append(" ), ");
+            }
+          }
+        }
+      }
+    }
+
+    if (warnIfFound && !configGroupMap.isEmpty()) {
+      output.replace(output.lastIndexOf(","), output.length(), "]");
+      warning("You have config groups present in the database with no " +
+        "corresponding service found, {}. Run --auto-fix-database to fix " +
+          "this automatically. Please backup Ambari Server database before running --auto-fix-database.", output.toString());
+    }
+
+    return configGroupMap;
+  }
```

Comment:
Looks like this case is not handled:
- A config group has its service name in the tag field and the service is deleted. Neither checkConfigGroupsHasServiceName() nor checkConfigGroupsForDeletedServices() detects it.

Reply:
I had two config groups before upgrade from 2.5:
I upgraded to 2.6 than 2.7 and deleted the service ambari_infra_solr. The configgroup table looks like:

group_name|tag|service_name
----------|---|--------------
cfg1|ZOOKEEPER|(null)	
cfg_infra|AMBARI_INFRA_SOLR|(null)	

When stating ambari server both methods warns:

    WARN - You have config groups present in the database with no service name, [(ConfigGroup) => ( cfg1 ), ( cfg_infra )]
    WARN - You have config groups present in the database with no corresponding service found, [(ConfigGroup, Service) => ( cfg1, null ), ( cfg_infra, null )]

After restarting ambari-server using `--auto-fix-database`

group_name|tag|service_name
----------|---|--------------
cfg1|ZOOKEEPER|ZOOKEEPER	
cfg_infra|AMBARI_INFRA_SOLR|(null)	

    WARN - You have config groups present in the database with no service name, [(ConfigGroup) => ( cfg_infra )]
    WARN - You have config groups present in the database with no corresponding service found, [(ConfigGroup, Service) => ( cfg_infra, null )]